## 1) Accuracy with pandas Series + accuracy_score

In [11]:
import pandas as pd
from sklearn.metrics import accuracy_score

true_classes = pd.Series([1, 0, 1, 1, 2, 0, 2, 1, 0, 1], name="true")
pred_classes = pd.Series([1, 0, 0, 1, 2, 0, 1, 2, 0, 1], name="pred")

acc = accuracy_score(true_classes, pred_classes)
print("Accuracy:", acc)

Accuracy: 0.7


## 2) Accuracy from first principles (manual)

In [12]:
true_list = [1, 0, 1, 1, 2, 0, 2, 1, 0, 1]
pred_list = [1, 0, 0, 1, 2, 0, 1, 2, 0, 1]

correct = sum(t == p for t, p in zip(true_list, pred_list))
total = len(true_list)
acc_manual = correct / total

print("Correct:", correct)
print("Total:", total)
print("Accuracy (manual):", acc_manual)

Correct: 7
Total: 10
Accuracy (manual): 0.7


## 3a) Precision using precision_score

In [13]:
from sklearn.metrics import precision_score

y_true = [1, 0, 1, 1, 0, 0, 1, 0, 1, 0]
y_pred = [1, 0, 1, 0, 0, 1, 1, 0, 0, 0]

prec = precision_score(y_true, y_pred)  # pos_label=1 by default
print("Precision:", prec)

Precision: 0.75


## 3b) Precision using the precision formula

In [14]:
y_true = [1, 0, 1, 1, 0, 0, 1, 0, 1, 0]
y_pred = [1, 0, 1, 0, 0, 1, 1, 0, 0, 0]

TP = sum((t == 1 and p == 1) for t, p in zip(y_true, y_pred))
FP = sum((t == 0 and p == 1) for t, p in zip(y_true, y_pred))

precision_manual = TP / (TP + FP) if (TP + FP) != 0 else 0

print("TP:", TP, "FP:", FP)
print("Precision (manual):", precision_manual)

TP: 3 FP: 1
Precision (manual): 0.75


## 3c) Why prioritize precision over accuracy here?

Accuracy can be misleading, especially if the dataset is imbalanced (many more non-cancer than cancer). 

A model could get high accuracy by predicting “non-cancer” most of the time.

Precision answers: “When the model says cancer, how often is it correct?”

In healthcare, false positives can cause stress, extra tests, cost, 

and sometimes invasive procedures, so you often care that positive predictions are trustworthy.

## 4) Recall (manual formula + recall_score)

In [15]:
from sklearn.metrics import recall_score

y_true = [1, 0, 1, 1, 0, 0, 1, 0, 1, 0]
y_pred = [1, 0, 1, 0, 0, 1, 1, 0, 0, 0]

# sklearn
rec = recall_score(y_true, y_pred)
print("Recall (sklearn):", rec)

# manual
TP = sum((t == 1 and p == 1) for t, p in zip(y_true, y_pred))
FN = sum((t == 1 and p == 0) for t, p in zip(y_true, y_pred))
recall_manual = TP / (TP + FN) if (TP + FN) != 0 else 0

print("TP:", TP, "FN:", FN)
print("Recall (manual):", recall_manual)

Recall (sklearn): 0.6
TP: 3 FN: 2
Recall (manual): 0.6


## 5a
Precision = 0.75 → When the model predicts “cancer”, 
it’s correct 75% of the time. (Still 25% false alarms.)

Recall = 0.60 → The model only catches 60% of actual cancer cases 
→ 40% are missed (false negatives), which is risky.

So this model is more conservative about calling cancer, but it misses too many real cancer cases.

## 5b
Usually for cancer screening, recall (sensitivity) is more important because:
A false negative (missed cancer) can delay treatment and lead to worse outcomes.